### ISSUE TITLE
Creating long versions of Albert

### ISSUE THREAD

**Issue by ontocord** [allenai/longformer#117]
Hi, I'm trying to following convert_model_to_long script to convert Albert to a long version by replacing the attention with LongformerSelfAttention. https://colab.research.google.com/drive/18ECsAeWFEsNxQ8C-8Q_OMbsfH0eC0sCJ?usp=sharing

I think I've got it working with just dropping in the attention and some architecture changes..  I haven't pretrained yet,  but cannot be sure as the BCP on eval is quite high ~ 17.  But the BCP is quite high for regular Albert as well (but strangely the BCP of AlbertLong is lower than regular Albert which is ~ 19)

One specific question I have is to confirm that what is returned by the LongformerSelfAttention is 3d tensor: batch x sequence length x hidden_dim, correct? 

Albert's q,k,v calculation produces a 4d tensor before doing a strange application against a dense tensor (768x768) to get the 3d tensor, and I am not sure if I'm applying LongformerSelfAttention correctly in this case.

 `context_layer`` = context_layer.permute(0, 2, 1, 3).contiguous()

        # Should find a better way to do this
        w = (
            self.dense.weight.t()
            .view(self.num_attention_heads, self.attention_head_size, self.hidden_size)
            .to(context_layer.dtype)
        )
        b = self.dense.bias.to(context_layer.dtype)

        projected_context_layer = torch.einsum("bfnd,ndh->bfh", context_layer, w) + b
        projected_context_layer_dropout = self.dropout(projected_context_layer)
        layernormed_context_layer = self.LayerNorm(input_ids + projected_context_layer_dropout)
        return (layernormed_context_layer, attention_probs) if output_attentions else (layernormed_context_layer,)

``

I modified to do this instead:

``
       attention_output = self.attention(hidden_states, attention_mask, output_attentions) #  head_mask,
        context_layer = attention_output[0]

        # this is from the albert implementation. Is this equivlaent to the below??
        # Should find a better way to do this
        #w = (
        #    self.dense.weight.t()
        #    .view(self.num_attention_heads, self.attention_head_size, self.hidden_size)
        #    .to(context_layer.dtype)
        #)
        #b = self.dense.bias.to(context_layer.dtype)
        
        new_shape = context_layer.size()[:-1] + (-1,)
        context_layer = context_layer.view(*new_shape)
        context_layer = self.dense(context_layer) # torch.einsum("bfnd,ndh->bfh", context_layer, w) + b
        context_layer = self.dropout(context_layer)
        context_layer = self.LayerNorm(hidden_states + context_layer) # projected_context_layer_dropout)
...

Any comments and suggestions is much appreciated.

**Comment by ibeltagy** [comment]
This is great. Thanks, @ontocord for giving this a try. 

> bpc for albert and albert-long

I agree that bpc~=19 for albert is quite high, and I don't know why this is the case, but the fact that you get bpc~=17 using albert-long (lower is better) means you are doing the conversation correctly. That said, I think it is worth trying to understand why albert bpc is quite high.

> projected_context_layer

I didn't try to run your code, but your code looks correct to me. The weird looking code is just a linear layer as you did. Also, the reasonable bpc score you are getting indicates that your code is correct. As a sanity check, you can mess up the conversation code and confirm that the bpc score increases.

**Comment by ontocord** [comment]
Thanks @ibeltagy for responding (and the great longformer code!). I will fiddle around to see why Albert's BPC is so high. It's good to know the BPC gets lower with using the longformer code. Will be back to report any results!

**Comment by ontocord** [comment]
Just to let you know, I've found that the larger Albert models produces more reasonable results. The smaller Albert models will not readily predict the correct token output which I think explains the lower BPC
